lib/datastore/datastore_container.es6.js

```diff
@@ -0,0 +1,205 @@
+
+  classes: [
+    {
+      name: 'HTTPRequest',
+      extends: 'foam.net.BaseHTTPRequest',
+
+      documentation: `Wrapper class for defining HTTPRequest delegate chain
+          under this controller. The desired chain is as follows:
+
+          RetryHTTPRequest
+              -> (agent-contextualized) AuthHTTPRequest
+              -> (default) HTTPRequest.
+
+          When no authAgent is specified, the AuthHTTPRequest is left out.
+
+          This ensures that requests that fail at the service level will be
+          retried a reasonable number of times, and that all requests requiring
+          credentials are appropriately authorized.`,
+
+      requires: [
+        'foam.net.BaseHTTPRequest',
+        'foam.net.RetryHTTPRequest',
+      ],
+      imports: [ 'authAgent?' ],
+
+      properties: [
+        {
+          class: 'Proxy',
+          of: 'foam.net.BaseHTTPRequest',
+          name: 'delegate',
+          factory: function() {
+            var BaseHTTPRequest = this.BaseHTTPRequest;
+            var RetryHTTPRequest = this.RetryHTTPRequest;
+
+            var baseProps = {};
+            this.BaseHTTPRequest.getAxiomsByClass(foam.core.Property)
+                .forEach(prop => baseProps[prop.name] = this[prop.name]);
+
+            // No auth agent => retry decorator only.
+            if (!this.authAgent) {
+              return RetryHTTPRequest.create(Object.assign({}, baseProps, {
+                delegate: BaseHTTPRequest.create(baseProps, this),
+              }), this);
+            }
+
+            var AuthHTTPRequest =
+                  this.authAgent.__subContext__.lookup('foam.net.HTTPRequest');
+            return RetryHTTPRequest.create(Object.assign({}, baseProps, {
+              delegate: AuthHTTPRequest.create(Object.assign({}, baseProps, {
+                delegate: BaseHTTPRequest.create(baseProps, this),
+              }), this),
+            }), this);
+          }
+        },
+      ],
+
+      methods: [
+        function init() {
+          this.validate();
+          this.SUPER();
+        },
+      ],
+    }
+  ],
+
+  properties: [
+    {
+      class: 'String',
+      documentation: `Email account used to authenticate against Google Cloud
+          services.`,
+      name: 'gcloudAuthEmail',
+      required: true
+    },
+    {
+      class: 'String',
+      documentation: `Private key used to authenticate against Google Cloud
+          services.`,
+      name: 'gcloudAuthPrivateKey',
+      required: true
+    },
+    {
+      class: 'String',
+      documentation: 'Google Cloud Project ID for Datastore deployment.',
+      name: 'gcloudProjectId',
+      value: 'web-confluence',
+    },
+    {
+      name: 'ctx',
+      documentation: 'Context in which components are created',
+      factory: function() {
+        // Cascade exports from logger and authAgent into context.
+        var ctx = this.__subContext__.createSubContext(this.logger);
+        if (!this.authAgent) return ctx;
```

Comment:
Might be a little better to invert this to do the extra initialization inside the if's body.